analytics/tools/patch_logfile.ipynb

```diff
@@ -0,0 +1,191 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from modyn.supervisor.internal.pipeline_executor.models import PipelineLogs\n",
+    "\n",
+    "\n",
+    "from modyn.supervisor.internal.grpc.enums import PipelineStage\n",
+    "# fill missing times in cumulative plot\n",
+    "\n",
+    "\n",
+    "from analytics.app.data.transform import logs_dataframe\n",
+    "from pathlib import Path\n",
+    "\n",
+    "\n",
+    "%load_ext autoreload\n",
+    "%autoreload 2"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Data loading"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# VARIABLES\n",
+    "\n",
+    "pipeline_logfile = Path(\"/Users/robinholzinger/robin/dev/eth/modyn/.data/evaluation_results/pipeline_5/pipeline.log\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "logs = PipelineLogs.model_validate_json(pipeline_logfile.read_text())\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "trains = [(l_ for l_ in logs.supervisor_logs.stage_runs if l_.id == PipelineStage.HANDLE_SINGLE_TRIGGER.name)]\n",
+    "evals = [(l_ for l_ in logs.supervisor_logs.stage_runs if l_.id == PipelineStage.EVALUATE_SINGLE.name and l_.info.eval_request.dataset_id == \"cglm_landmark_min25-test\")]\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from analytics.app.data.transform import dfs_models_and_evals\n",
+    "\n",
+    "df_logs = logs_dataframe(logs)\n",
+    "# max_timestamp = df_logs[\"sample_time\"].max()\n",
+    "max_timestamp = 1703682949  # TODO: code needs minor change so that logfile contain max timestamp\n",
```

Comment:
there is an open todo, although I used this notebook to test it and it worked for me :)

Reply:
Thanks for the pointer, the comment is outdated, originally I thought it made sense to a dedicated "max_timestamp" log metric to the logfile. When we update the sample_time in the `pipeline_executor` after executing a `HANDLE_NEW_DATA` loop, we can achieve the same by querying the max sample_time in the log dataframe.